text/go/0001-go-bom.md

```diff
@@ -0,0 +1,163 @@
+
+* Requires {`go-dist`} during `build`
+* Provides none
+
+#### Build
+
+The build phase will perform a few tasks as mentioned above.
+
+1. Perform the BOM dependency tool retrieval from the dep-server. It will get added to a build-time layer. It will
+   contribute a BOM entry for the tool itself on the `build.toml` `[bom]` section.
+
+2. Execute BOM generation with the installed tool via the `cyclonedx-gomod app -o bom.json` command in the root
+   directory of the application. There is an unresolved question around determining which main package to run against if
+   the application source contains multiple `main` packages.
+
+3. Unmarshal the CycloneDX JSON BOM and transform each BOM entry into a
+   [`postal.Dependency`](https://github.com/paketo-buildpacks/packit/blob/c5a40518f2c6bd913ade999b9e2d58d6892d2ea9/postal/buildpack.go#L12)
+   type. This step and step 5 will eventually go away if we support CycloneDX as a BOM format.
+
+4. (Optionally) generate `CPE` and `pURL` from the data and add it to the
+   `postal.Dependency`.
+
+5. Run the
+   [`postal.GenerateBillOfMaterials`](https://github.com/paketo-buildpacks/packit/blob/c5a40518f2c6bd913ade999b9e2d58d6892d2ea9/postal/service.go#L186)
+   command.
+
+6. Clean up the generated BOM file
+
+#### Potential Future Optimizations
+
+Given future user interest in a way to opt out of this buildpack beyond a custom order grouping, an optimization that
+could be made is the inclusion of a
+`BP_ENABLE_MODULE_BOM` environment variable that can be set during container build-time.
+
+## Rationale
+
+Performing BOM generation with the CycloneDX Go Module tool has some advantages. First, it minimizes the extra domain
+knowledge we would have to develop and maintain a program to recurse through all the go modules and their dependencies.
+
+Second, there is a high likelihood that we will support CycloneDX as a BOM format in the future. In that case, we can
+simply run the tool and return that BOM without any additional formatting or conversion. Going this route will minimize
+the change needed down the road, and we can get a CycloneDX output we can feel confident about.
+
+## Alternatives
+
+An alternative to the option to using the CycloneDX BOM tool would be to hand-create the BOM as described. We could go
+through each go module directory, pull information out of each `go.mod` file and then build out our own BOM with that
+data. This is a viable option if installing a tool is not a reasonable choice.
+
+An alternative to performing all of these steps in a separate buildpack would be to add the CycloneDX tool installation
+as a step in the Go Build buildpack, and then run BOM creation alone in a buildpack before the start commands. This is
+less than ideal because it's not transparent or intuitive to install it during the go specific build process. It makes
+more sense to separate all module BOM creation logic into its own buildpack, so that users may easily opt out.
```

Comment:
The more I read of the RFC, the more I'm asking myself why we aren't pursuing this alternative. In this paragraph it appears that we have 2 concerns: 1) we need to install a tool, and 2) we want users to be able to opt-out.

For 1, I don't see why we couldn't introduce a buildpack that just does the tool installation and makes it available during build.
For 2, I also don't see why we need a separate buildpack to make this possible.

Can we elaborate a bit more on these concerns?

Reply:
This is fair Ryan. Part of the decision behind the pattern we used in the Node Module BOM was because while we could've done it the way you've described in point # 1, except for the fact we'd still need a separate buildpack for the Node Start buildpack case. Because of that outlier, we decided to go with the pattern we have now to be consistent across the language family.

I don't think that same concern is relevant here, since the BOM only needs to be created for one order group scenario. Therefore, the plan you've described is viable as far as I know

Reply:
I think I am much less concerned about the duplication of logic than I am about the decoupling of the BOM from the artifact and also the implicit coupling of this BOM buildpack to the implementations of those buildpacks doing the "work" of the build process.

Additionally, recording the BOM into a single place makes it more awkward for us to support other BOM use-cases going forward. For example, I'd like to see us also record parts of the BOM that are only used during the build phase. This separate buildpack will need to do quite a bit of reverse-engineering to discover how a particular artifact was used in the build process. This is obviously more difficult than just having each buildpack report what it used directly.